needless_borrow false positive: "use of moved value"

### Summary

Beginning in nightly-2022-10-25, the needless_borrow lint makes suggestions that involve changing a borrow to a move, even when a move would not be legal.

### Lint Name

needless_borrow

### Reproducer

```rust
fn main() {
    let string = String::new();
    for _i in 0..10 {
        f(string);
    }
}

fn f<T: AsRef<str>>(_: T) {}
```

```console
$ cargo +nightly-2022-10-24 clippy
    Checking testing v0.0.0 (/git/testing)
    Finished dev [unoptimized + debuginfo] target(s) in 0.03s

$ cargo +nightly-2022-10-25 clippy
    Checking testing v0.0.0 (/git/testing)
warning: the borrowed expression implements the required traits
 --> src/main.rs:4:11
  |
4 |         f(&string);
  |           ^^^^^^^ help: change this to: `string`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_borrow
  = note: `#[warn(clippy::needless_borrow)]` on by default

warning: `testing` (bin "testing") generated 1 warning
    Finished dev [unoptimized + debuginfo] target(s) in 0.20s
```

After making the suggested change:

```console
$ cargo check
    Checking testing v0.0.0 (/git/testing)
error[E0382]: use of moved value: `string`
 --> src/main.rs:4:11
  |
2 |     let string = String::new();
  |         ------ move occurs because `string` has type `String`, which does not implement the `Copy` trait
3 |     for _i in 0..10 {
4 |         f(string);
  |           ^^^^^^ value moved here, in previous iteration of loop

For more information about this error, try `rustc --explain E0382`.
error: could not compile `testing` due to previous error
```

### Version

```text
rustc 1.66.0-nightly (758f19645 2022-10-24)
binary: rustc
commit-hash: 758f19645b8ebce61ea52d1f6672fd057bc8dbee
commit-date: 2022-10-24
host: x86_64-unknown-linux-gnu
release: 1.66.0-nightly
LLVM version: 15.0.2
```


### Additional Labels

@rustbot label +I-suggestion-causes-error

--- a/clippy_lints/src/dereference.rs
+++ b/clippy_lints/src/dereference.rs
@@ -9,6 +9,7 @@ use clippy_utils::{
 };
 use rustc_ast::util::parser::{PREC_POSTFIX, PREC_PREFIX};
 use rustc_data_structures::fx::FxIndexMap;
+use rustc_data_structures::graph::iterate::{CycleDetector, TriColorDepthFirstSearch};
 use rustc_errors::Applicability;
 use rustc_hir::intravisit::{walk_ty, Visitor};
 use rustc_hir::{
@@ -1242,6 +1243,8 @@ fn referent_used_exactly_once<'a, 'tcx>(
         && let Some(statement) = mir.basic_blocks[location.block].statements.get(location.statement_index)
         && let StatementKind::Assign(box (_, Rvalue::Ref(_, _, place))) = statement.kind
         && !place.has_deref()
+        // Ensure not in a loop (https://github.com/rust-lang/rust-clippy/issues/9710)
+        && TriColorDepthFirstSearch::new(&mir.basic_blocks).run_from(location.block, &mut CycleDetector).is_none()
     {
         let body_owner_local_def_id = cx.tcx.hir().enclosing_body_owner(reference.hir_id);
         if possible_borrowers

--- a/tests/ui/needless_borrow.fixed
+++ b/tests/ui/needless_borrow.fixed
@@ -408,3 +408,15 @@ mod issue_9111 {
         a.extend(&[]); // vs a.extend([]);
     }
 }
+
+#[allow(dead_code)]
+mod issue_9710 {
+    fn main() {
+        let string = String::new();
+        for _i in 0..10 {
+            f(&string);
+        }
+    }
+
+    fn f<T: AsRef<str>>(_: T) {}
+}

--- a/tests/ui/needless_borrow.rs
+++ b/tests/ui/needless_borrow.rs
@@ -408,3 +408,15 @@ mod issue_9111 {
         a.extend(&[]); // vs a.extend([]);
     }
 }
+
+#[allow(dead_code)]
+mod issue_9710 {
+    fn main() {
+        let string = String::new();
+        for _i in 0..10 {
+            f(&string);
+        }
+    }
+
+    fn f<T: AsRef<str>>(_: T) {}
+}

